File: src/terminal/terminal.zig
const std = @import("std");

pub const TerminalSize = struct {
    cols: u16,
    rows: u16,
    width_px: u16,
    height_px: u16,
};

/// Global flag for SIGWINCH (terminal resize) detection
/// Using atomic to safely communicate between signal handler and main thread
var resize_pending = std.atomic.Value(bool).init(false);
var interrupt_pending = std.atomic.Value(bool).init(false);

/// Global state for emergency cleanup on signal
var global_original_termios: ?std.posix.termios = null;
var global_stdin_fd: std.posix.fd_t = std.posix.STDIN_FILENO;

/// SIGWINCH signal handler - sets the resize flag
fn handleSigwinch(_: c_int) callconv(.c) void {
    resize_pending.store(true, .release);
}

/// Signal handler for SIGINT/SIGTERM
/// Sets a flag so the main loop can exit cleanly
fn handleTermSignal(_: c_int) callconv(.c) void {
    if (interrupt_pending.swap(true, .acq_rel)) {
        // Force exit on second signal
        std.process.exit(1);
    }
}

pub const Terminal = struct {
    stdin_fd: std.posix.fd_t,
    original_termios: ?std.posix.termios,
    mouse_enabled: bool,
    sigwinch_installed: bool,

    pub fn init() Terminal {
        return .{
            .stdin_fd = std.posix.STDIN_FILENO,
            .original_termios = null,
            .mouse_enabled = false,
            .sigwinch_installed = false,
        };
    }

    /// Install SIGWINCH handler for terminal resize detection
    pub fn installResizeHandler(self: *Terminal) !void {
        if (self.sigwinch_installed) return;

        const act = std.posix.Sigaction{
            .handler = .{ .handler = handleSigwinch },
            .mask = std.posix.sigemptyset(),
            .flags = 0,
        };

        std.posix.sigaction(std.posix.SIG.WINCH, &act, null);
        self.sigwinch_installed = true;
    }

    /// Check if terminal was resized (clears the flag)
    pub fn checkResize(_: *Terminal) bool {
        return resize_pending.swap(false, .acq_rel);
    }

    /// Check if an interrupt signal was received
    pub fn checkInterrupt(_: *Terminal) bool {
        return interrupt_pending.load(.acquire);
    }

    /// Enter raw mode (disable line buffering, echo)
    pub fn enterRawMode(self: *Terminal) !void {
        // Check if stdin is a TTY
        if (!std.posix.isatty(self.stdin_fd)) {
            std.debug.print("Error: stdin is not a terminal\n", .{});
            std.debug.print("termweb requires an interactive terminal (TTY) to run.\n", .{});
            return error.NotATty;
        }

        // Save original settings
        self.original_termios = try std.posix.tcgetattr(self.stdin_fd);

        // Save global state for emergency signal cleanup
        global_original_termios = self.original_termios;
        global_stdin_fd = self.stdin_fd;

        // Install emergency cleanup handlers for SIGINT/SIGTERM
        const term_act = std.posix.Sigaction{
            .handler = .{ .handler = handleTermSignal },
            .mask = std.posix.sigemptyset(),
            .flags = 0,
        };
        std.posix.sigaction(std.posix.SIG.INT, &term_act, null);
        std.posix.sigaction(std.posix.SIG.TERM, &term_act, null);

        var raw = self.original_termios.?;

        // Disable canonical mode, echo, signals
        raw.lflag.ICANON = false;
        raw.lflag.ECHO = false;
        raw.lflag.ISIG = false;
        raw.lflag.IEXTEN = false;

        // Disable input processing
        raw.iflag.IXON = false;
        raw.iflag.ICRNL = false;
        raw.iflag.BRKINT = false;
        raw.iflag.INPCK = false;
        raw.iflag.ISTRIP = false;

        // Disable output processing
        raw.oflag.OPOST = false;

        // Character-at-a-time input (non-blocking)
        raw.cc[@intFromEnum(std.posix.V.MIN)] = 0;
        raw.cc[@intFromEnum(std.posix.V.TIME)] = 0;

        try std.posix.tcsetattr(self.stdin_fd, .FLUSH, raw);
    }

    /// Enable Kitty mouse protocol with pixel coordinates
    pub fn enableMouse(self: *Terminal) !void {
        if (self.mouse_enabled) return;

        var stdout_buf: [256]u8 = undefined;
        const stdout_file = std.fs.File.stdout();
        var stdout_writer = stdout_file.writer(&stdout_buf);
        const writer = &stdout_writer.interface;

        // Enable mouse tracking (all events: press, release, move)
        // 1003h = all motion tracking (needed for cursor display)
        try writer.writeAll("\x1b[?1003h");

        // Enable SGR extended mouse mode (for better coordinate parsing)
        try writer.writeAll("\x1b[?1006h");

        // Enable SGR pixel mouse mode (actual pixel coordinates, not cells)
        try writer.writeAll("\x1b[?1016h");

        // Enable Kitty keyboard protocol (flags: 1=disambiguate, 8=report all keys)
        // Flag 8 is needed for Cmd+letter keys to be reported instead of intercepted
        try writer.writeAll("\x1b[>9u");

        // Enable bracketed paste mode (for proper paste handling)
        try writer.writeAll("\x1b[?2004h");

        try writer.flush();
        self.mouse_enabled = true;
    }

    /// Disable mouse protocol
    pub fn disableMouse(self: *Terminal) !void {
        if (!self.mouse_enabled) return;

        var stdout_buf: [256]u8 = undefined;
        const stdout_file = std.fs.File.stdout();
        var stdout_writer = stdout_file.writer(&stdout_buf);
        const writer = &stdout_writer.interface;

        // Disable Kitty keyboard protocol
        try writer.writeAll("\x1b[<u");

        // Disable bracketed paste mode
        try writer.writeAll("\x1b[?2004l");

        // Disable SGR pixel mouse mode
        try writer.writeAll("\x1b[?1016l");

        // Disable mouse tracking
        try writer.writeAll("\x1b[?1003l");

        // Disable SGR mouse mode
        try writer.writeAll("\x1b[?1006l");

        try writer.flush();
        self.mouse_enabled = false;
    }

    /// Restore original terminal settings
    pub fn restore(self: *Terminal) !void {
        // Disable mouse first (while still in raw mode)
        self.disableMouse() catch {};

        // Drain any buffered input (mouse events, etc.)
        self.drainInput();

        // Restore original terminal settings
        if (self.original_termios) |orig| {
            try std.posix.tcsetattr(self.stdin_fd, .FLUSH, orig);
        }

        // Clear global state (prevents signal handler from double-restoring)
        global_original_termios = null;
    }

    /// Drain any pending input from stdin
    fn drainInput(self: *Terminal) void {
        var buf: [256]u8 = undefined;
        // O_NONBLOCK value (0x0004 on macOS, 0x800 on Linux)
        const O_NONBLOCK: usize = if (@import("builtin").os.tag == .macos) 0x0004 else 0x800;

        // Set non-blocking read temporarily
        const flags = std.posix.fcntl(self.stdin_fd, std.posix.F.GETFL, 0) catch return;
        _ = std.posix.fcntl(self.stdin_fd, std.posix.F.SETFL, flags | O_NONBLOCK) catch return;
        defer _ = std.posix.fcntl(self.stdin_fd, std.posix.F.SETFL, flags) catch {};

        // Read and discard all pending input - loop multiple times to catch any stragglers
        var attempts: u32 = 0;
        while (attempts < 10) : (attempts += 1) {
            var drained_any = false;
            while (true) {
                const n = std.posix.read(self.stdin_fd, &buf) catch break;
                if (n == 0) break;
                drained_any = true;
            }
            if (!drained_any) break;
            // Small delay to let more input arrive
            std.Thread.sleep(5 * std.time.ns_per_ms);
        }
    }

    /// Get terminal size with pixel dimensions
    pub fn getSize(self: *Terminal) !TerminalSize {
        var ws: std.posix.winsize = undefined;

        const result = std.c.ioctl(self.stdin_fd, std.posix.T.IOCGWINSZ, &ws);

        if (result != 0) {
            return error.IoctlFailed;
        }

        return TerminalSize{
            .cols = ws.col,
            .rows = ws.row,
            .width_px = ws.xpixel,
            .height_px = ws.ypixel,
        };
    }

    /// Cleanup (called on deinit or crash)
    pub fn deinit(self: *Terminal) void {
        self.restore() catch {};
    }
};

File: lib/terminal.js
/**
 * Terminal utilities - raw mode, mouse, size, etc.
 */

const tty = require('tty');
const readline = require('readline');

// ANSI escape codes
const ESC = '\x1b';
const CSI = `${ESC}[`;

class Terminal {
  constructor() {
    this._rawMode = false;
    this._mouseEnabled = false;
    this._originalMode = null;
  }

  /**
   * Enter raw mode (no echo, no line buffering)
   */
  enterRawMode() {
    if (this._rawMode) return;

    if (process.stdin.isTTY) {
      this._originalMode = process.stdin.isRaw;
      process.stdin.setRawMode(true);
    }

    this._rawMode = true;
  }

  /**
   * Exit raw mode
   */
  exitRawMode() {
    if (!this._rawMode) return;

    if (process.stdin.isTTY && this._originalMode !== null) {
      process.stdin.setRawMode(this._originalMode);
    }

    this._rawMode = false;
  }

  /**
   * Enable mouse tracking
   * @param {Object} options
   * @param {boolean} [options.sgr] - Use SGR extended mode
   * @param {boolean} [options.pixels] - Use pixel coordinates
   */
  enableMouse(options = {}) {
    if (this._mouseEnabled) return;

    // Enable mouse tracking
    process.stdout.write(`${CSI}?1000h`); // Basic mouse
    process.stdout.write(`${CSI}?1002h`); // Button events
    process.stdout.write(`${CSI}?1003h`); // All motion events

    if (options.sgr !== false) {
      process.stdout.write(`${CSI}?1006h`); // SGR extended mode
    }

    if (options.pixels) {
      process.stdout.write(`${CSI}?1016h`); // Pixel coordinates
    }

    this._mouseEnabled = true;
  }

  /**
   * Disable mouse tracking
   */
  disableMouse() {
    if (!this._mouseEnabled) return;

    process.stdout.write(`${CSI}?1016l`);
    process.stdout.write(`${CSI}?1006l`);
    process.stdout.write(`${CSI}?1003l`);
    process.stdout.write(`${CSI}?1002l`);
    process.stdout.write(`${CSI}?1000l`);

    this._mouseEnabled = false;
  }

  /**
   * Get terminal size
   * @returns {{ cols: number, rows: number, width: number, height: number }}
   */
  getSize() {
    const cols = process.stdout.columns || 80;
    const rows = process.stdout.rows || 24;

    // Try to get pixel size via ioctl
    let width = cols * 10; // Fallback
    let height = rows * 20;

    // Use TIOCGWINSZ if available
    try {
      const size = process.stdout.getWindowSize?.();
      if (size) {
        width = size[0] || width;
        height = size[1] || height;
      }
    } catch (e) {
      // Ignore
    }

    return { cols, rows, width, height };
  }

  /**
   * Clear the screen
   */
  clear() {
    process.stdout.write(`${CSI}2J`);
    process.stdout.write(`${CSI}H`);
  }

  /**
   * Move cursor to position
   * @param {number} row - 1-indexed row
   * @param {number} col - 1-indexed column
   */
  moveCursor(row, col) {
    process.stdout.write(`${CSI}${row};${col}H`);
  }

  /**
   * Hide cursor
   */
  hideCursor() {
    process.stdout.write(`${CSI}?25l`);
  }

  /**
   * Show cursor
   */
  showCursor() {
    process.stdout.write(`${CSI}?25h`);
  }

  /**
   * Set terminal title
   * @param {string} title
   */
  setTitle(title) {
    process.stdout.write(`${ESC}]0;${title}\x07`);
  }

  /**
   * Enable alternate screen buffer
   */
  enterAltScreen() {
    process.stdout.write(`${CSI}?1049h`);
  }

  /**
   * Exit alternate screen buffer
   */
  exitAltScreen() {
    process.stdout.write(`${CSI}?1049l`);
  }

  /**
   * Read keyboard input
   * @returns {AsyncGenerator<{ key: string, ctrl: boolean, meta: boolean, shift: boolean }>}
   */
  async *readKeys() {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      terminal: true,
    });

    readline.emitKeypressEvents(process.stdin, rl);

    if (process.stdin.isTTY) {
      process.stdin.setRawMode(true);
    }

    for await (const [str, key] of this._keyEvents()) {
      yield {
        key: key?.name || str,
        sequence: str,
        ctrl: key?.ctrl || false,
        meta: key?.meta || false,
        shift: key?.shift || false,
      };
    }
  }

  /**
   * Internal key event generator
   * @private
   */
  async *_keyEvents() {
    const events = [];
    let resolve;

    const handler = (str, key) => {
      if (resolve) {
        resolve([str, key]);
        resolve = null;
      } else {
        events.push([str, key]);
      }
    };

    process.stdin.on('keypress', handler);

    try {
      while (true) {
        if (events.length > 0) {
          yield events.shift();
        } else {
          yield await new Promise((r) => (resolve = r));
        }
      }
    } finally {
      process.stdin.off('keypress', handler);
    }
  }

  /**
   * Cleanup - restore terminal state
   */
  cleanup() {
    this.disableMouse();
    this.exitRawMode();
    this.showCursor();
    this.exitAltScreen();
    process.stdout.write(`${CSI}0m`); // Reset attributes
  }
}

module.exports = Terminal;
